cache/demux.go

```diff
@@ -167,16 +177,32 @@ func (d *demux) updateStoreLocked(events []*clientv3.Event) {
 	d.maxRev = events[len(events)-1].Kv.ModRevision
 }
 
-func (d *demux) broadcastLocked(events []*clientv3.Event) {
-	firstRev := events[0].Kv.ModRevision
-	lastRev := events[len(events)-1].Kv.ModRevision
+func (d *demux) broadcastLocked(resp clientv3.WatchResponse) {
+	if resp.IsProgressNotify() {
+		for w, nextRev := range d.activeWatchers {
+			if nextRev >= resp.Header.Revision {
+				continue
+			}
+			if !w.enqueueResponse(clientv3.WatchResponse{
+				Header: etcdserverpb.ResponseHeader{
+					Revision: resp.Header.Revision,
+				},
+			}) { // overflow → lagging
+				d.laggingWatchers[w] = nextRev
+				delete(d.activeWatchers, w)
+			} else {
+				d.activeWatchers[w] = resp.Header.Revision + 1
+			}
+		}
+		return
+	}
+	if len(resp.Events) == 0 {
+		return
+	}
 
+	events := resp.Events
+	lastRev := events[len(events)-1].Kv.ModRevision
 	for w, nextRev := range d.activeWatchers {
-		if nextRev != 0 && firstRev > nextRev {
```

Comment:
I don't remember why I removed it, but now I'm thinking it might not be safe to do so. At least without more consideration.